### ISSUE TITLE
Rendering issues with webgl and ligatures

### ISSUE THREAD

**Issue by LabhanshAgrawal** [xtermjs/xterm.js#3303]
## Details
- Browser and browser version: Chrome 89
- OS version: macOS catalina
- xterm.js version: At commit 376b29673ba174934b1b6339ef3eed8449fec529

### Steps to reproduce

Happened to me using the demo with zsh with powerlevel10k
Comparing canvas vs webgl with ligatures

Canvas
<img width="300" alt="Screenshot 2021-04-08 at 20 17 55" src="https://user-images.githubusercontent.com/16598275/114049413-0ffbfe80-98a9-11eb-977d-b58b335e435a.png">
Webgl
<img width="300" alt="Screenshot 2021-04-08 at 20 19 35" src="https://user-images.githubusercontent.com/16598275/114049425-11c5c200-98a9-11eb-8bc2-d831815ce62b.png">

Below, I'm using suggestions and syntax highlighting plugins for zsh
It's suggesting === (untyped part of the suggestion should be in gray)

On typing =
Both
<img width="300" alt="Screenshot 2021-04-08 at 20 18 09" src="https://user-images.githubusercontent.com/16598275/114051008-5e5dcd00-98aa-11eb-8575-1a7a22b11bb1.png">

On typing ==
Canvas
<img width="300" alt="Screenshot 2021-04-08 at 20 19 09" src="https://user-images.githubusercontent.com/16598275/114049430-125e5880-98a9-11eb-9ef3-127ca6e9914a.png">
Webgl
<img width="300" alt="Screenshot 2021-04-08 at 20 20 21" src="https://user-images.githubusercontent.com/16598275/114049432-12f6ef00-98a9-11eb-8d72-110976075ac1.png">

On typing ===
Canvas
<img width="300" alt="Screenshot 2021-04-08 at 20 19 18" src="https://user-images.githubusercontent.com/16598275/114049435-12f6ef00-98a9-11eb-8d21-1892367656a6.png">
Webgl
<img width="300" alt="Screenshot 2021-04-08 at 20 20 28" src="https://user-images.githubusercontent.com/16598275/114049438-138f8580-98a9-11eb-95aa-1363fd6b8b67.png">

Looks like remnants of earlier paint are there in webgl

**Comment by Tyriar** [comment]
The grey `=` is tracked in https://github.com/xtermjs/xterm.js/issues/3288, not sure what's happening with the `:`, what font is this?

**Comment by LabhanshAgrawal** [comment]
Fira Code
It's the demo running without any changes (just turned on ligatures & webgl and opened zsh)

**Comment by LabhanshAgrawal** [comment]
It looks like the character from inside the cursor is staying on.

**Comment by Tyriar** [comment]
> It looks like the character from inside the cursor is staying on.

I think that's actually ideal imo, but I don't use ligatures so not 100% sure. The dom renderer is a little bad atm since the cursor highlights all chars in the ligature.

**Comment by LabhanshAgrawal** [comment]
what I meant was that the artifact for character inside the cursor is left even after typing and the cursor moving 1 place ahead
If you look at the sequence from 1->3 ='s
the grey `=` that stays on is the one from inside the cursor

For the dom I guess it's happening since the joined characters are in the same `span` and the cursor is covering it full. Might be hard to fix as it'd need to be a separate element then. And might cause issues with emoji's or cjk as they are multi character too.

**Comment by jerch** [comment]
> For the dom I guess it's happening since the joined characters are in the same span and the cursor is covering it full. Might be hard to fix as it'd need to be a separate element then. And might cause issues with emoji's or cjk as they are multi character too.

Yep, compound unicode chars (grapheme clusters) basically have the same issue, imho this was the reason for inventing IMEs in the first place - the user needs a way to interact with its "construction phase" (cursor access to its atomic entities).

AFAIK word processors treat clusters this way - once they are constructed the normal cursor would not allow to jump into the atomics anymore (treated as a single char, cursor jumps right over it, BS would delete it as a whole), but have the ability to re-enter the construction phase by some additional action to modify the atomics again.

While we can learn from word processors, how a proper cursor based interaction with clusters may look like, it is not quite easy to apply for a grid-aligned terminal. We basically would need a way to re-enter the IME construction once again, no clue if thats possible somehow.

Ligatures as used from coding fonts are somewhat special here as they output different glyphs for consecutive chars, that still show normal monospaced cursor progression. Imho here we should break the ligature, maybe simply skip the joiner if the cursor is above?

**Comment by LabhanshAgrawal** [comment]
Looks like the rendering issues go further
See https://github.com/vercel/hyper/issues/5757

**Comment by npezza93** [comment]
fwiw @Tyriar @LabhanshAgrawal  if you manually clear the texture atlas this goes away. perhaps theres some key that's missing in cache key?

**Comment by jerch** [comment]
@npezza93 Does this still happen with v5.3? Because in older versions there was a bug in the cache altas, which got fixed around 5.3 (see #4534).

**Comment by npezza93** [comment]
@jerch yep!
